Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -499,20 +419,75 @@ extension SwiftLanguageServer {
     req[keys.request] = self.requests.editor_close
     req[keys.name] = uri.pseudoPath
 
-    // Clear settings that should not be cached for closed documents.
-    self.currentDiagnostics[uri] = nil
-
     _ = try? self.sourcekitd.sendSync(req)
 
     await semanticTokensManager.discardSemanticTokens(for: note.textDocument.uri)
   }
 
+  /// Cancels any in-flight tasks to send a `PublishedDiagnosticsNotification` after edits.
+  ///
+  /// - Important: Cancelling in-flight publish diagnostics tasks is required before any edit. See the comment inside
+  ///   `documentDiagnostic` for more details.
+  private func cancelInFlightPublishDiagnosticsTask(for document: DocumentURI) {
+    if let inFlightTask = inFlightPublishDiagnosticsTasks[document] {
+      inFlightTask.cancel()
+    }
+  }
+
+  /// If the client doesn't support pull diagnostics, compute diagnostics for the latest version of the given document
+  /// and send a `PublishDiagnosticsNotification` to the client for it.
+  private func publishDiagnosticsIfNeeded(for document: DocumentURI) {
+    guard enablePublishDiagnostics else {
+      return
+    }
+    guard diagnosticsEnabled(for: document) else {
+      log("Ignoring diagnostics for blacklisted file \(document.pseudoPath)", level: .debug)
```

Comment:
`blacklisted` -> `blocklisted`: Am I missing something here. I think `blacklisted` is the correct word, right?

> You could log this as `.info` during an open IMO (rather than spam the debug log here).

I’d prefer to keep it as-is for now. That way we have less state in the log and the log message can’t leave the log window.

Also, `debug` is a very low log level that is usually only captured if `sourcekit-lsp` is being debugged, so it shouldn’t spam anything bey default.

Reply:
> Am I missing something here. I think blacklisted is the correct word, right?

Blocklist is the inclusive alternative.

> Also, debug is a very low log level that is usually only captured if sourcekit-lsp is being debugged, so it shouldn’t spam anything bey default.

Sure, it just seems like something that would be useful to know *without* having debugging enabled. And having it once at document open seems reasonable to me - it's enabled/disabled for the entire document anyway.

> That way we have less state in the log and the log message can’t leave the log window.

What extra state would there be here?

Reply:
> Sure, it just seems like something that would be useful to know without having debugging enabled. And having it once at document open seems reasonable to me - it's enabled/disabled for the entire document anyway.

Honestly, I think I’ll just remove the log message completely. The check for `diagnosticsEnabled` is so damn simple (ignore URIs that start with `git://` or `hg://` for which we don’t even make sense to get diagnostics for.

> What extra state would there be here?

Essentially the state that a comment would say: Don’t log again, we already logged that we’re not emitting diagnostics when opening the file (which might get out of sync).